### DISCUSSION TITLE
building fails on nixos

### DISCUSSION THREAD

**Discussion by wizardwatch** [qtile/qtile#2602] [Q&A]
Qtile's build on NixOS fails due to it not passing 21 test cases. https://github.com/NixOS/nixpkgs/pull/123257 , preventing the use of 0.17.0 and stopping me from trying master with wayland. The author of the linked pr seems to have identified the issue as being https://github.com/qtile/qtile/issues/1634 although I am not sure how they identified it nor have I tried fixing it yet (I am not well versed in qtiles code base). Are there any suggestions on how to debug this? I have exact same log as seen in the pr above. Do you want me to file an issue or perhaps send something similar in the mailing list?

**Comment by m-col** [comment]
I don't think #1634 is the issue, I think it is dependencies missing from the test environment, i.e. https://github.com/qtile/qtile/issues/2466

Some of the errors in the test log output from NixOS/nixpkgs#123257:

>            cairocffi.pixbuf.ImageLoadingError: Pixbuf error: Unrecognized image file format

This causes some of the problems, and happens when an SVG image is trying to be loaded. Perhaps `librsvg` needs to be installed manually in the test env?

Then in other tests, some test windows are not appearing, and this message is printed to stderr:

>       Fontconfig error: Cannot load default config file

so `fontconfig` might need to be installed, or perhaps something else related to it, though i'm not sure why it would be needed for these test windows.

Later, there is this error:

>                FileNotFoundError: [Errno 2] No such file or directory: '/build/source/test/../bin/qtile'

The tests are assuming they are being run within the git repository, as this path is leaving the `test` folder and entering the `bin` folder to find a script. Perhaps install the entire `qtile` folder to the test environment and run it from within `test/`?

>        /build/source/libqtile/utils.py:158: UnixCommandNotFound: [Errno 2] No such file or directory: 'sensors'

This needs the `lm_sensors` (Arch name) package.

**Reply by wizardwatch** [reply]
Thanks for the response. I believe I have identified why https://github.com/qtile/qtile/issues/1634 was called out. I am hitting https://gitlab.freedesktop.org/xorg/xserver/-/issues/1177 and the issues give similar errors.

**Reply by wizardwatch** [reply]
I have not been able to fix the fontconfig issue. Is the fontconfig path hardcoded some where?

**Reply by wizardwatch** [reply]
On a different front, I added lm_sensors and it is now detected, but I am getting a similar error to https://github.com/qtile/qtile/issues/1922 . My log at this point looks like this [log.txt](https://github.com/qtile/qtile/files/6759086/log.txt)

**Reply by wizardwatch** [reply]
> I have not been able to fix the fontconfig issue. Is the fontconfig path hardcoded some where?

I have made progress on this front. I added a file to the build with a basic font config and set the env var FONTCONFIG_FILE to that file. My log now looks like this. I still don't pass the tests but any progress is good progress. 
[qtile_log.txt](https://github.com/qtile/qtile/files/6794411/qtile_log.txt)

**Reply by m-col** [reply]
Nice. For the `sensors` issue you could tell pytest to ignore that test `--ignore=test/<path>`

**Reply by wizardwatch** [reply]
Patching shebangs let me cut one more error. test_qtile_cmd now passes!

**Reply by wizardwatch** [reply]
I am starting to think the fontconfig errors might not be causing the failure. Any other things anyone think it might be? Current log 
[qtile-log2.txt](https://github.com/qtile/qtile/files/6796156/qtile-log2.txt)

**Reply by wizardwatch** [reply]
Comparing my logs to those from the Github action I see the Github action never has any of the unhandled event errors. They are seemingly suppressed as the tests that pass also have them, but finding the cause may reveal the issue.

**Reply by m-col** [reply]
Those unhandled event messages are log output that are also logged during tests that pass, it is just that pytest only shows you the stdout of tests that fail. I don't think these are problematic.

If I'm reading correctly, the remaining errors are all ones related to `manager.create_window`. The window never seems to appear - and this is true whether it is a window created manually with xcffib or by spawning `xcalc` or similar. This makes me suspect something is up with the general environment. Is it possible to run the tests manually on the same server?

**Reply by wizardwatch** [reply]
> Those unhandled event messages are log output that are also logged during tests that pass, it is just that pytest only shows you the stdout of tests that fail. I don't think these are problematic.
> 
> If I'm reading correctly, the remaining errors are all ones related to `manager.create_window`. The window never seems to appear - and this is true whether it is a window created manually with xcffib or by spawning `xcalc` or similar. This makes me suspect something is up with the general environment. Is it possible to run the tests manually on the same server?

I rewrote some stuff and am now sort of able to. The good news is, in a sandboxed shell all the tests pass! The bad news is it seems like the issue is with how sandboxed builds interact with Xephyr so it will probably never pass all of the tests during build. I am going to make a new pr with what ever tests failed disabled so that I can start focusing on pywlroots bring up.
